flax/struct.py

```diff
@@ -70,66 +70,74 @@ def __apply__(self, *args):
-    updates = {}
-    for name in data_fields:
-      if name not in state:
-        raise ValueError(f'Missing field {name} in state dict while restoring'
-                         f' an instance of {clz.__name__}')
-      value = getattr(x, name)
-      value_state = state.pop(name)
-      updates[name] = serialization.from_state_dict(value, value_state)
-    if state:
-      names = ','.join(state.keys())
-      raise ValueError(f'Unknown field(s) "{names}" in state dict while'
-                       f' restoring an instance of {clz.__name__}')
-    return x.replace(**updates)
-
-  serialization.register_serialization_state(
-      data_clz, to_state_dict, from_state_dict)
-
-  return data_clz
+  def wrapped(clz):
+    data_clz = dataclasses.dataclass(frozen=frozen)(clz)
+    meta_fields = []
+    data_fields = []
+    for name, field_info in data_clz.__dataclass_fields__.items():
+      is_pytree_node = field_info.metadata.get('pytree_node', True)
+      if is_pytree_node:
+        data_fields.append(name)
+      else:
+        meta_fields.append(name)
+
+    def replace(self, **updates):
+      """"Returns a new object replacing the specified fields with new values."""
+      return dataclasses.replace(self, **updates)
+
+    data_clz.replace = replace
+
+    def iterate_clz(x):
+      meta = tuple(getattr(x, name) for name in meta_fields)
+      data = tuple(getattr(x, name) for name in data_fields)
+      return data, meta
+
+    def clz_from_iterable(meta, data):
+      meta_args = tuple(zip(meta_fields, meta))
+      data_args = tuple(zip(data_fields, data))
+      kwargs = dict(meta_args + data_args)
+      return data_clz(**kwargs)
+
+    jax.tree_util.register_pytree_node(data_clz,
+                                      iterate_clz,
+                                      clz_from_iterable)
+
+    def to_state_dict(x):
+      state_dict = {name: serialization.to_state_dict(getattr(x, name))
+                    for name in data_fields}
+      return state_dict
+
+    def from_state_dict(x, state):
+      """Restore the state of a data class."""
+      state = state.copy()  # copy the state so we can pop the restored fields.
+      updates = {}
+      for name in data_fields:
+        if name not in state:
+          raise ValueError(f'Missing field {name} in state dict while restoring'
+                          f' an instance of {clz.__name__}')
+        value = getattr(x, name)
+        value_state = state.pop(name)
+        updates[name] = serialization.from_state_dict(value, value_state)
+      if state:
+        names = ','.join(state.keys())
+        raise ValueError(f'Unknown field(s) "{names}" in state dict while'
+                        f' restoring an instance of {clz.__name__}')
+      return x.replace(**updates)
+
+    serialization.register_serialization_state(
+        data_clz, to_state_dict, from_state_dict)
+
+    return data_clz
+  if clz is None:
```

Comment:
sorry could you explain what this new conditional on clz is for?  I thought this was a required argument?

Reply:
It's just to let `dataclass` be called with or without the `frozen` argument when used as a decorator. eg,

```
@dataclass(frozen=False)
class X:
  ...
```

desugars to `X=dataclass(frozen=True)(X)` so `clz` is `None` in that call to `dataclass`. Meanwhile 

```
@dataclass
class X:
   ...
```

desguars to `X=dataclass(X)`, and there `clz` is `X`.

This is the same pattern used the original dataclasses from https://github.com/ericvsmith/dataclasses/blob/master/dataclasses.py#L935.